---
layout: post
category: "other"
title:  "Soft article in WeChat"
tags: [other]
---

### RNA biology

---------------------------

**Mitchell Guttman**, California Institute of Technology, [labpage](http://guttmanlab.caltech.edu/index.php)

* RAP-DNA, RAP-RNA
* The presence of thousands of functional large non-coding RNAs (lncRNAs) in the mammalian genome represents a missing component in our understanding of genome regulation.
* Our lab aims to understand how lncRNAs control gene expression programs and cell state decisions in the context of mouse embryonic stem cells. We are an integrated team of experimental and computational biologists who work together to address these questions using genomic approaches in conjunction with biochemistry, molecular biology, cell biology, and computational biology.

---------------------------

**Xiangdong Fu**, UCSD, [labpage](http://cmm.ucsd.edu/fu/)

* eCLIP, GRID-seq
* The Fu laboratory is interested in molecular and cell biology of RNA metabolism and regulation in higher eukaryotic cells. Current research interests in the Fu lab include the regulation of alternative splicing, functional RNA elements in mammalian genomes, transcription/splicing coupling, nuclear architecture, and cellular reprogramming.

---------------------------

**Howard Chang**, Stanford University, [labpage](http://changlab.stanford.edu/)

* lncRNA, ATAC-seq, icSHAPE, PARIS
* The Chang lab is focused on how the activities of hundreds or even thousands of genes (gene parties) are coordinated to achieve biological meaning. We have pioneered methods to predict, dissect, and control large-scale gene regulatory programs; these methods have provided insights into human development, cancer, and aging. A particular interest is how cells know and remember their locations in the body, particularly with the help of long noncoding RNAs.

---------------------------

### RNA secondary structure

---------------------------

**Alain Laederach**, University of North Carolina, Chapel Hill, [labpage](https://ribosnitch.bio.unc.edu/research/)

* RibosNitch, SNPfold
* The Laederach Lab is interested in better understanding the relationship between RNA structure and folding and human disease. We use a combination of computational and experimental approaches to study the process of RNA folding and in the cells. In particular, we develop novel approaches to analyze and interpret chemical and enzymatic mapping data on a genomic scale. We aim to fundamentally understand the role of RNA structure in controlling post-transcriptional regulatory mechanisms, and to interpret structure as a secondary layer of information ([Nature comment, 2014](http://www.nature.com/nature/journal/v505/n7485/full/505621a.html)).  We are particularly interested in how human genetic variation affects RNA regulatory structure. We investigate the relationship between disease-associated Single Nucleotide Polymorphisms occurring in Human UTRs and their effect on RNA structure to determine if they form a RiboSNitch.

---------------------------

**Kevin Weeks**, University of North Carolina, Chapel Hill, [labpage](https://weeks.chem.unc.edu/)

* QuShape, Differential SHAPE, SHAPE-MaP, RING-MaP, deltaSHAPE, ShapeMapper 2
* Chemical Microscopes for High-Content RNA Structure Analysis
* Structure and Function in the Transcriptome

---------------------------

**Anna Marie Pyle**, Yale University, HHMI, [labpage](https://pylelab.org/)

* HOTAIR secondary structure, HCV secondary structure
* We explore RNA Biology, studying the molecular interactions important for RNA structures and the activities of RNA-protein machines.  Using tools that range from crystallography to cell culture, we seek to understand the impact of RNA architecture and dynamics on the life of the cell. 

---------------------------

**Walter N. Moss**, Iowa State University, [labpage](http://www.mosslab.org/)

* RNA2DMut, RNAStructuromeDB
* validate viral ncRNA's predicted structures, identify interacting molecules, determine their localization within the cell and determine their effects on host cells.
* Viral cis-regulatory elements
* Host-virus interactions

---------------------------

**Janusz M. Bujnicki**, International Institute of Molecular and Cell Biology in Warsaw, Poland, [labpage](http://genesilico.pl/)

* SimRNA: RNA 3D structure modeling (participate [RNA Puzzles](http://www.rnapuzzles.org/))
* Our group is involved in theoretical and experimental research on nucleic acids and proteins. The current focus is on RNA sequence-structure-function relationships (in particular 3D modeling), RNA-protein complexes, and enzymes acting on RNA.

---------------------------

**Silvi Rouskin**, MIT, [labpage](https://www.rouskinlab.com/)

* DMS-seq & DMS-MaPseq
* Separating alternative structures formed from the same underlying sequence (RNA Structure Control of Alternative Splicing)

---------------------------

**Dan Herschlag**, STANFORD UNIVERSITY, [labpage](http://herschlaglab.stanford.edu/)

* RNA folding, RNA catalysis, protein catalysis, in vivo RNA structure and interactions
* We are particularly interested in questions of how enzymes work, how RNA folds, how proteins recognize RNA, and the roles of RNA/protein interactions in regulation and control, and the evolution of molecules and molecular interactions.

---------------------------

**Sharon Aviran**, UC Davis, [labpage](https://aviranlab.bme.ucdavis.edu/)

* dStruct, patteRNA, RNAprob
* RNA structure and dynamics
* Our lab develops novel computational methods for inferring RNA dynamics from experiments and theory, with applications ranging from basic research to biomolecular engineering and synthetic biology.

---------------------------

**David Mathews**, University of Rochester, [labpage](https://rna.urmc.rochester.edu/index.html)

* RNAstructure
* Our goal is to automate the modeling of RNA structure and function from genome sequence to 3D structure.

---------------------------

**Julius B. Lucks**, Northwestern University, [labpage](http://luckslab.org/)

* SHAPE-Seq
* Pushing the Limits of RNA Design with Cellular Engineering
* Next Generation RNA Structure Characterization
* Harvesting RNA Design Principles from Nature: RNA-protein interaction

---------------------------

### RNA-RNA interaction

---------------------------

**Zhipeng Lu**, University of Southern California School of Pharmacy, [labpage](https://zhipenglulab.org/)

* PARIS
* Our research is focused on “RNA machines” in living cells. We develop and apply novel technologies to understand the structures and functions of RNA molecules in basic cellular processes, with the ultimate goal of treating human diseases, including genetic disorders, cancers and viral infections.

---------------------------

**Irmtraud Margret Meyer**, Max-Delbrück-Centrum für Molekulare Medizin, [labpage](https://www.mdc-berlin.de/meyer)

* RNA structure and trans RNA-RNA interaction (recently review chapter: [Deciphering the Universe of RNA Structures and trans RNA–RNA Interactions of Transcriptomes In Vivo: From Experimental Protocols to Computational Analyses](https://link.springer.com/chapter/10.1007/978-3-319-92967-5_9))
* Computational methods development (NAR, [A comprehensive comparison of general RNA-RNA interaction prediction methods](https://www.ncbi.nlm.nih.gov/pubmed/26673718))
* The computational methods we devise employ machine learning techniques that are capable of detecting even subtle sequence (and other) signals in high-throughput transcriptome data. We typically employ fully probabilistic methods that enable us to assign reliability values to our predictions.
* On the biological side, we are studying transcriptome regulation in a variety of exciting biological systems in vivo ranging from early human embryogenesis to neurogenesis in the fruit fly to how pathogens interact with different hosts. To this end, we closely collaborate with several experimental groups on and off campus. 

---------------------------


### RNA-protein interaction

---------------------------

**Gene Yeo**, UCSD, [labpage](https://yeolab.github.io/)

* eCLIP, CLIPPER
* how RNA binding proteins and RNA modifications affect cellular homeostasis in human pluripotent stem cells.
* how defects in RNA binding proteins cause neurological disease, such as ALS.
* post-transcriptional processing of RNAs by multiple mechanisms.
* develop new computational approaches to decipher biological meaning from thousands of single-cell RNA-seq/proteomics data.

---------------------------

### RNA-DNA interaction

---------------------------

**Sheng Zhong**, UCSD, [labpage](http://systemsbio.ucsd.edu/index.html)

* We study gene regulation and cellular behavior by developing statistical and experimental methods. Our primary goal is to develop new technologies to map molecular networks, including RNA-RNA interactome [[MARIO, Nat Comm, 2016]](http://www.nature.com/articles/ncomms12023), RNA-chromatin interactome [[MARGI, Curr Biol, 2017]](http://www.cell.com/current-biology/fulltext/S0960-9822(17)30011-8), and protein-protein interactome. Our secondary quest is to model the variations of these networks in three axes, namely developmental time, personal difference, and evolutionary change. Our major tools include epigenomic and single-cell assays, single-molecule imaging, statistical modeling, and large scale computation.

---------------------------

### Stress granule

---------------------------

**Roy Parker**, UNIVERSITY OF COLORADO BOULDER, [labpage](https://www.colorado.edu/lab/parkergroup/)

* Our goal is to understand the molecular mechanisms that control mRNA stability and translation rate in eukaryotic cells, using yeast as a model system.

---------------------------

**Jernej Ule**, UCL/Crick institute, [labpage](http://www.ulelab.info/index.php)

* HiCLIP, iCLIP, RNP granule
* The goal of our research group is to reveal **how RNPs regulate the life cycle of mRNAs** in neurons, and how this can go wrong in neurologic diseases. To study the assembly of RNPs, we obtain detailed maps of protein-RNA binding sites by using transcriptomic techniques. For this purpose, we developed the nucleotide-resolution UV crosslinking and immunoprecipitation (iCLIP), which identifies protein-RNA contacts by using a series of steps, as described in the figure above. We are further developing similar methods, as well as computational tools to interpret the high-throughput sequencing data. Thereby, we gain a comprehensive view of RNP assembly and dynamics within intact cells.

---------------------------

